### ISSUE TITLE
FA3/4 does not guard against NaN in unused V cache entries

### ISSUE THREAD

**Issue by liangel-02** [Dao-AILab/flash-attention#2374]
when using `seqused_k` with a KV cache, entries beyond `seqused_k` can contain NaNs. in FA2, the invalid entries gets set to 0 (see [here](https://github.com/Dao-AILab/flash-attention/blob/5f682309c0d0fe1446cf030096fc1784d3dde20a/csrc/flash_attn/src/flash_fwd_kernel.h#L313))

however, FA3/FA4 load the full cache via TMA, not the per-sequence `seqused_k`. the extra K values are correctly masked out but when loading in V, it includes NaN values in unused cache slots. the attention score masking correctly produces P = 0 for positions beyond `seqused_k`, but the PV matmul then computes `0 * NaN = NaN`. 

see repro below for FA4

```python 
import torch
from flash_attn.cute import flash_attn_varlen_func


def repro():
    torch.manual_seed(42)
    device = "cuda"
    dtype = torch.bfloat16

    batch_size = 4
    num_heads = 8
    head_dim = 64
    cache_size = 128
    actual_kv_lens = [32, 64, 96, 48]

    # Single-token queries (decode-like)
    total_q = batch_size
    q = torch.randn(total_q, num_heads, head_dim, device=device, dtype=dtype)
    cu_seqlens_q = torch.arange(0, batch_size + 1, device=device, dtype=torch.int32)

    # Build KV cache: valid data for [0, seqused_k), NaN for [seqused_k, cache_size)
    k_cache = torch.full(
        (batch_size * cache_size, num_heads, head_dim),
        float("nan"), device=device, dtype=dtype,
    )
    v_cache = torch.full(
        (batch_size * cache_size, num_heads, head_dim),
        float("nan"), device=device, dtype=dtype,
    )
    for i in range(batch_size):
        offset = i * cache_size
        k_cache[offset : offset + actual_kv_lens[i]] = torch.randn(
            actual_kv_lens[i], num_heads, head_dim, device=device, dtype=dtype
        )
        v_cache[offset : offset + actual_kv_lens[i]] = torch.randn(
            actual_kv_lens[i], num_heads, head_dim, device=device, dtype=dtype
        )

    cu_seqlens_k = torch.arange(
        0, (batch_size + 1) * cache_size, cache_size,
        device=device, dtype=torch.int32,
    )
    seqused_k = torch.tensor(actual_kv_lens, device=device, dtype=torch.int32)

    # ---- Reference: tightly packed KV, no unused slots ----
    k_seqs = [k_cache[i * cache_size : i * cache_size + actual_kv_lens[i]] for i in range(batch_size)]
    v_seqs = [v_cache[i * cache_size : i * cache_size + actual_kv_lens[i]] for i in range(batch_size)]
    k_packed = torch.cat(k_seqs)
    v_packed = torch.cat(v_seqs)
    cu_seqlens_k_ref = torch.zeros(batch_size + 1, device=device, dtype=torch.int32)
    cu_seqlens_k_ref[1:] = torch.tensor(actual_kv_lens, device=device).cumsum(0).int()

    out_ref = flash_attn_varlen_func(
        q, k_packed, v_packed,
        cu_seqlens_q=cu_seqlens_q,
        cu_seqlens_k=cu_seqlens_k_ref,
        max_seqlen_q=1,
        max_seqlen_k=max(actual_kv_lens),
    )[0]

    # ---- Bug case: NaN in unused cache slots + seqused_k ----
    out_nan = flash_attn_varlen_func(
        q, k_cache, v_cache,
        cu_seqlens_q=cu_seqlens_q,
        cu_seqlens_k=cu_seqlens_k,
        max_seqlen_q=1,
        max_seqlen_k=cache_size,
        seqused_k=seqused_k,
    )[0]

    # ---- Results ----
    print("=" * 60)
    print("seqused_k NaN handling repro")
    print("=" * 60)

    nan_count = out_nan.isnan().sum().item()
    total_elems = out_nan.numel()

    print(f"\n1. Reference (tightly packed, no cache):  OK")
    print(f"2. Cache with NaN fill + seqused_k:       {nan_count}/{total_elems} elements are NaN")

    print(f"\nPer-sequence breakdown:")
    for i in range(batch_size):
        is_nan = out_nan[i].isnan().all().item()
        ref_ok = not out_ref[i].isnan().any().item()
        print(f"  batch {i}: seqused_k={actual_kv_lens[i]:3d}, "
              f"cache_size={cache_size}, "
              f"unused_slots={cache_size - actual_kv_lens[i]:3d}, "
              f"output_all_nan={is_nan}, ref_ok={ref_ok}")


if __name__ == "__main__":
    repro()
```

cc @drisspg

**Comment by NJX-njx** [comment]
I'd like to work on this. I've traced the root cause in FA4 (\): the last partial K/V block loads V via TMA for the full block width, including positions ≥ \ that may contain NaN. The QK mask correctly zeros those positions in P (via -inf → softmax → 0), but the subsequent PV GEMM computes 0×NaN = NaN per IEEE 754, propagating NaN to the output. Planning to fix by predicate-filling the out-of-range V positions to 0.0 before the PV GEMM in the last partial block.
